include/boost/url/decoded_view.hpp

```diff
@@ -0,0 +1,1049 @@
+
+    using traits_type = std::char_traits<char>;
+
+public:
+    /** Type of a decoded character
+    */
+    using value_type = char;
+
+    /// @copydoc value_type
+    using reference = char;
+
+    /// @copydoc value_type
+    using const_reference = char;
+
+    /** The unsigned integer type used to represent size.
+    */
+    using size_type = std::size_t;
+
+    /** The signed integer type used to represent differences.
+    */
+    using difference_type = std::ptrdiff_t;
+
+    /** A read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+#ifdef BOOST_URL_DOCS
+    using const_iterator = __see_below__;
+#else
+    class const_iterator;
+#endif
+
+    /// @copydoc const_iterator
+    using iterator = const_iterator;
+
+    /** A reverse read-only bidirectional iterator to the decoded range.
+
+        This is a read-only bidirectional iterator to
+        the decoded characters.
+
+     */
+    using reverse_iterator = std::reverse_iterator<iterator>;
+
+    /// @copydoc reverse_iterator
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+    /** Constructor
+
+        Default-constructed objects represent
+        the empty string.
+
+     */
+    BOOST_CONSTEXPR
+    decoded_view(
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(nullptr)
+        , n_(0)
+        , dn_(0)
+        , plus_to_space_(opt.plus_to_space)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str), opt)
+    {}
+
+    /** Constructor
+     */
+    decoded_view(
+        char const* str,
+        size_type len,
+        pct_decode_opts const& opt = {}) noexcept
+        : decoded_view(string_view(str, len), opt)
+    {}
+
+    /** Constructor
+     */
+#ifndef BOOST_URL_DOCS
+    template <
+        class End,
+        typename std::enable_if<std::is_same<End, char const*>::value, int>::type = 0>
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        End last,
+        pct_decode_opts const& opt = {}) noexcept
+#else
+    BOOST_CXX14_CONSTEXPR
+    decoded_view(
+        char const* first,
+        char const* last,
+        pct_decode_opts const& opt = {}) noexcept
+#endif
+        : decoded_view(string_view(first, last), opt)
+    {
+        BOOST_ASSERT(last - first >= 0);
+    }
+
+    /** Constructor
+     */
+    decoded_view(
+        urls::string_view str,
+        pct_decode_opts const& opt = {}) noexcept
+        : p_(str.data())
+        , n_(str.size())
+        , dn_(pct_decode_bytes_unchecked(str))
+        , plus_to_space_(opt.plus_to_space)
+    {
+        error_code ec;
+        auto is_unambiguous = [](unsigned char ch)
+        {
+            return ch != '%';
+        };
+        validate_pct_encoding(str, ec, opt, is_unambiguous);
```

Comment:
This `is_unambiguous` thing doesn't sound right. You probably want to set `opt.non_normal_is_error = false` and leave out the CharSet parameter. See:
https://github.com/CPPAlliance/url/blob/ef2db3c1e52474b881812da92ae495cd936ed150/include/boost/url/pct_encoding.hpp#L100